File: app/src/main/java/com/example/android/popularmovies/ui/fragment/MovieDetailFragment.java
package com.example.android.popularmovies.ui.fragment;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.app.ShareCompat;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import com.example.android.popularmovies.R;
import com.example.android.popularmovies.model.MovieModel;
import com.example.android.popularmovies.model.ReviewModel;
import com.example.android.popularmovies.model.TrailerModel;
import com.example.android.popularmovies.presenter.MovieDetailPresenter;
import com.example.android.popularmovies.ui.view.MovieDetailView;
import com.example.android.popularmovies.util.MovieUtil;
import com.squareup.picasso.Picasso;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import dagger.android.support.AndroidSupportInjection;

/**
 * Created by msk-1196 on 7/15/17.
 */

public class MovieDetailFragment extends Fragment implements MovieDetailView {
    public static final String CONTENT_KEY = "content.movie";
    private static final String SCROLL_STATE_KEY = "com.example.popularmovies.scrollstate";
    private static final String REVIEWS_STATE_KEY = "com.example.popularmovies.reviewsstate";
    private static final String VIDEOS_STATE_KEY = "com.example.popularmovies.videosstate";

    @BindView(R.id.scrollView) ScrollView mScrollView;
    @BindView(R.id.backdrop) ImageView mBackdropImageView;
    @BindView(R.id.movie_poster) ImageView mPosterImageView;
    @BindView(R.id.title) TextView mTitleTextView;
    @BindView(R.id.release_date) TextView mReleaseDateTextView;
    @BindView(R.id.rate_avg) TextView mRatingAverageTextView;
    @BindView(R.id.rate_count) TextView mRatingCountTextView;
    @BindView(R.id.overview) TextView mOverviewTextView;
    @BindView(R.id.fab_favorite) FloatingActionButton mFavoriteFab;
    @BindView(R.id.video_container) LinearLayout mVideoContainer;
    @BindView(R.id.review_container) LinearLayout mReviewContainer;

    @Inject MovieDetailPresenter mPresenter;
    private int[] scrollState = new int[2];

    public static MovieDetailFragment newInstance(MovieModel movie){
        MovieDetailFragment fragment = new MovieDetailFragment();
        Bundle args = new Bundle();
        args.putParcelable(CONTENT_KEY, movie);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (savedInstanceState != null){
            scrollState = savedInstanceState.getIntArray(SCROLL_STATE_KEY);
            mPresenter.setVideos(savedInstanceState.getParcelableArrayList(VIDEOS_STATE_KEY));
            mPresenter.setReviews(savedInstanceState.getParcelableArrayList(REVIEWS_STATE_KEY));
        }

        mPresenter.loadTrailer();
        mPresenter.loadReview();
        if (scrollState != null && scrollState.length == 2) {
            mScrollView.post(() ->
                    mScrollView.scrollTo(scrollState[0], scrollState[1]));
        }

        mPresenter.checkFavorite();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_movie_detail, container, false);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ButterKnife.bind(this, view);

        mFavoriteFab.setOnClickListener(view1 -> mPresenter.toggleFavorite());

        Bundle args = getArguments();
        if (args != null){
            mPresenter.setDetailForMovie(args.getParcelable(CONTENT_KEY));
        }
    }

    @Override
    public void showDetailMovie(MovieModel movie) {
        getActivity().setTitle(movie.getTitle());
        mTitleTextView.setText(movie.getTitle());
        mReleaseDateTextView.setText(movie.getReleaseDate());
        // show the vote_average
        mRatingAverageTextView.setText(String.format(Locale.getDefault(),
                "%.1f/10", movie.getVoteAverage()));
        mOverviewTextView.setText(movie.getOverview());
        mRatingCountTextView.setText(String.valueOf(movie.getVoteCount()));
        if (movie.isFavorite())
            mFavoriteFab.setImageResource(R.drawable.ic_favorite_fill);
        else
            mFavoriteFab.setImageResource(R.drawable.ic_favorite_border);

        if (movie.getBackdropPath() != null){
            Picasso.with(getActivity())
                    .load(MovieUtil.getFullBackdropPath(movie.getBackdropPath()))
                    .into(mBackdropImageView);
        }

        if (movie.getPosterPath() != null){
            Picasso.with(getActivity())
                    .load(MovieUtil.getFullPosterPath(movie.getPosterPath()))
                    .into(mPosterImageView);
        }
    }

    @Override
    public void showTrailer(List<TrailerModel> trailers) {
        if (trailers == null || trailers.size() == 0){
            mVideoContainer.setVisibility(View.GONE);
            return;
        }

        for (int i = 1; i < mVideoContainer.getChildCount(); ++i)
            mVideoContainer.removeViewAt(i);

        final LayoutInflater layoutInflater = LayoutInflater.from(getActivity());

        for (TrailerModel trailer : trailers){
            final View view = layoutInflater.inflate(R.layout.item_video_layout, mVideoContainer, false);
            final ImageView mThumbnailImageView = (ImageView) view.findViewById(R.id.thumbnail);
            if (trailer.getSite().equalsIgnoreCase("YouTube"))
                Picasso.with(getActivity())
                    .load(MovieUtil.getYoutubeThumbnailPath(trailer.getKey()))
                    .into(mThumbnailImageView);

            final TextView mVideoTitle = (TextView) view.findViewById(R.id.video_title);
            mVideoTitle.setText(String.format(Locale.getDefault(), "[%s] %s", trailer.getType(),
                    trailer.getName()));
            view.setTag(trailer);
            view.setOnClickListener(view1 -> {
                TrailerModel t = (TrailerModel) view1.getTag();
                if (t.getSite().equalsIgnoreCase("YouTube")){
                    Intent intent = new Intent(Intent.ACTION_VIEW, MovieUtil.getVideoUri(t.getKey()));
                    if (intent.resolveActivity(getActivity().getPackageManager()) != null)
                        startActivity(intent);
                } else {
                    Toast.makeText(getContext(), R.string.video_not_supported, Toast.LENGTH_SHORT).show();
                }
            });

            mVideoContainer.addView(view);

        }

        mVideoContainer.setVisibility(View.VISIBLE);
        getActivity().invalidateOptionsMenu();

    }

    @Override
    public void showReview(List<ReviewModel> reviews) {
        if (reviews == null || reviews.size() == 0){
            mReviewContainer.setVisibility(View.GONE);
            return;
        }

        for (int i = 1; i < mReviewContainer.getChildCount(); ++i)
            mReviewContainer.removeViewAt(i);

        final LayoutInflater layoutInflater = LayoutInflater.from(getActivity());

        for (ReviewModel review : reviews){
            final View view = layoutInflater.inflate(R.layout.item_review_layout, mReviewContainer, false);

            final TextView mAuthorTextView = (TextView) view.findViewById(R.id.review_author);
            mAuthorTextView.setText(MovieUtil.customNullAuthor(review.getAuthor()));

            final TextView mContentTextView = (TextView) view.findViewById(R.id.review_body);
            mContentTextView.setText(review.getContent());

            mReviewContainer.addView(view);

        }

        mReviewContainer.setVisibility(View.VISIBLE);
    }

    @Override
    public void toggleFavoriteFab(boolean favorite) {
        mFavoriteFab.setImageResource(favorite? R.drawable.ic_favorite_fill :
                R.drawable.ic_favorite_border);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        outState.putIntArray(SCROLL_STATE_KEY, new int[]{mScrollView.getScrollX(), mScrollView.getScrollY()});
        outState.putParcelableArrayList(VIDEOS_STATE_KEY, new ArrayList<>(mPresenter.getVideos()));
        outState.putParcelableArrayList(REVIEWS_STATE_KEY, new ArrayList<>(mPresenter.getReviews()));
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.detail_menu, menu);
    }

    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        if (mPresenter.getVideos() != null && mPresenter.getVideos().size() > 0)
            menu.findItem(R.id.action_share).setVisible(true);
        else
            menu.findItem(R.id.action_share).setVisible(false);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.action_share:
                share(mPresenter.getVideos().get(0));
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void share(TrailerModel trailerModel) {
        Intent intent = ShareCompat.IntentBuilder.from(getActivity())
                .setType("text/plain")
                .setText(getString(R.string.share_template, String.format(Locale.getDefault(),
                        "[%s] %s", trailerModel.getType(), trailerModel.getName()),
                        MovieUtil.getVideoUri(trailerModel.getKey()).toString()))
                .getIntent();
        startActivity(Intent.createChooser(intent, getString(R.string.action_share)));
    }
}

File: app/src/main/java/com/example/android/popularmovies/ui/fragment/MovieDetailFragment.kt
package com.example.android.popularmovies.ui.fragment

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.Fragment
import android.support.v4.app.ShareCompat
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.CompoundButton
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView
import android.widget.Toast

import com.example.android.popularmovies.R
import com.example.android.popularmovies.model.MovieModel
import com.example.android.popularmovies.model.ReviewModel
import com.example.android.popularmovies.model.TrailerModel
import com.example.android.popularmovies.presenter.MovieDetailPresenter
import com.example.android.popularmovies.ui.view.MovieDetailView
import com.example.android.popularmovies.util.MovieUtil
import com.squareup.picasso.Picasso

import java.util.ArrayList
import java.util.Locale

import javax.inject.Inject

import dagger.android.support.AndroidSupportInjection
import kotlinx.android.synthetic.main.fragment_movie_detail.*

/**
 * Created by msk-1196 on 7/15/17.
 */

class MovieDetailFragment : Fragment(), MovieDetailView {

    @Inject lateinit var mPresenter: MovieDetailPresenter
    private var scrollState: IntArray? = IntArray(2)

    override fun onAttach(context: Context?) {
        AndroidSupportInjection.inject(this)
        super.onAttach(context)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState != null) {
            scrollState = savedInstanceState.getIntArray(SCROLL_STATE_KEY)
            mPresenter!!.setVideos(savedInstanceState.getParcelableArrayList(VIDEOS_STATE_KEY)!!)
            mPresenter!!.setReviews(savedInstanceState.getParcelableArrayList(REVIEWS_STATE_KEY)!!)
        }

        mPresenter!!.loadTrailer()
        mPresenter!!.loadReview()
        if (scrollState != null && scrollState!!.size == 2) {
            scrollView.post { scrollView.scrollTo(scrollState!![0], scrollState!![1]) }
        }

        mPresenter!!.checkFavorite()
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater!!.inflate(R.layout.fragment_movie_detail, container, false)
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        fab_favorite.setOnClickListener { _ -> mPresenter!!.toggleFavorite() }

        val args = arguments
        if (args != null) {
            mPresenter!!.setDetailForMovie(args.getParcelable(CONTENT_KEY)!!)
        }
    }

    override fun showDetailMovie(movie: MovieModel) {
        activity.title = movie.title
        title.text = movie.title
        release_date.text = movie.releaseDate
        // show the vote_average
        rate_avg.text = String.format(Locale.getDefault(),
                "%.1f/10", movie.voteAverage)
        overview.text = movie.overview
        rate_count.text = movie.voteCount.toString()
        if (movie.isFavorite)
            fab_favorite.setImageResource(R.drawable.ic_favorite_fill)
        else
            fab_favorite.setImageResource(R.drawable.ic_favorite_border)

        if (movie.backdropPath != null) {
            Picasso.with(activity)
                    .load(MovieUtil.getFullBackdropPath(movie.backdropPath))
                    .into(backdrop)
        }

        if (movie.posterPath != null) {
            Picasso.with(activity)
                    .load(MovieUtil.getFullPosterPath(movie.posterPath))
                    .into(movie_poster)
        }
    }

    override fun showTrailer(trailers: List<TrailerModel>) {
        if (trailers.isEmpty()) {
            video_container.visibility = View.GONE
            return
        }

        for (i in 1 until video_container.childCount)
            video_container.removeViewAt(i)

        val layoutInflater = LayoutInflater.from(activity)

        for (trailer in trailers) {
            val view = layoutInflater.inflate(R.layout.item_video_layout,
                    video_container, false)
            val mThumbnailImageView = view.findViewById<View>(R.id.thumbnail) as ImageView
            if (trailer.site!!.equals("YouTube", ignoreCase = true))
                Picasso.with(activity)
                        .load(MovieUtil.getYoutubeThumbnailPath(trailer.key!!))
                        .into(mThumbnailImageView)

            val mVideoTitle = view.findViewById<View>(R.id.video_title) as TextView
            mVideoTitle.text = String.format(Locale.getDefault(), "[%s] %s", trailer.type,
                    trailer.name)
            view.tag = trailer
            view.setOnClickListener { view1 ->
                val t = view1.tag as TrailerModel
                if (t.site!!.equals("YouTube", ignoreCase = true)) {
                    val intent = Intent(Intent.ACTION_VIEW, MovieUtil.getVideoUri(t.key!!))
                    if (intent.resolveActivity(activity.packageManager) != null)
                        startActivity(intent)
                } else {
                    Toast.makeText(context, R.string.video_not_supported, Toast.LENGTH_SHORT).show()
                }
            }

            video_container.addView(view)

        }

        video_container.visibility = View.VISIBLE
        activity.invalidateOptionsMenu()

    }

    override fun showReview(reviews: List<ReviewModel>) {
        if (reviews.isEmpty()) {
            review_container.visibility = View.GONE
            return
        }

        for (i in 1 until review_container.childCount)
            review_container.removeViewAt(i)

        val layoutInflater = LayoutInflater.from(activity)

        for (review in reviews) {
            val view = layoutInflater.inflate(R.layout.item_review_layout,
                    review_container, false)

            val mAuthorTextView = view.findViewById<View>(R.id.review_author) as TextView
            mAuthorTextView.text = MovieUtil.customNullAuthor(review.author)

            val mContentTextView = view.findViewById<View>(R.id.review_body) as TextView
            mContentTextView.text = review.content

            review_container.addView(view)

        }

        review_container.visibility = View.VISIBLE
    }

    override fun toggleFavoriteFab(favorite: Boolean) {
        fab_favorite.setImageResource(if (favorite)
            R.drawable.ic_favorite_fill
        else
            R.drawable.ic_favorite_border)
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)

        outState!!.putIntArray(SCROLL_STATE_KEY, intArrayOf(scrollView.scrollX, scrollView.scrollY))
        outState.putParcelableArrayList(VIDEOS_STATE_KEY, ArrayList(mPresenter!!.videos!!))
        outState.putParcelableArrayList(REVIEWS_STATE_KEY, ArrayList(mPresenter!!.reviews!!))
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        inflater!!.inflate(R.menu.detail_menu, menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu?) {
        if (mPresenter!!.videos != null && mPresenter!!.videos!!.size > 0)
            menu!!.findItem(R.id.action_share).isVisible = true
        else
            menu!!.findItem(R.id.action_share).isVisible = false
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item!!.itemId) {
            R.id.action_share -> {
                share(mPresenter!!.videos!![0])
                return true
            }

            else -> return super.onOptionsItemSelected(item)
        }
    }

    private fun share(trailerModel: TrailerModel) {
        val intent = ShareCompat.IntentBuilder.from(activity)
                .setType("text/plain")
                .setText(getString(R.string.share_template, String.format(Locale.getDefault(),
                        "[%s] %s", trailerModel.type, trailerModel.name),
                        MovieUtil.getVideoUri(trailerModel.key!!).toString()))
                .intent
        startActivity(Intent.createChooser(intent, getString(R.string.action_share)))
    }

    companion object {
        const val CONTENT_KEY = "content.movie"
        private const val SCROLL_STATE_KEY = "com.example.popularmovies.scrollstate"
        private const val REVIEWS_STATE_KEY = "com.example.popularmovies.reviewsstate"
        private const val VIDEOS_STATE_KEY = "com.example.popularmovies.videosstate"

        fun newInstance(movie: MovieModel): MovieDetailFragment {
            val fragment = MovieDetailFragment()
            val args = Bundle()
            args.putParcelable(CONTENT_KEY, movie)
            fragment.arguments = args
            return fragment
        }
    }
}
